src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs

```diff
@@ -137,18 +140,32 @@ static void CacheMethod(Type handler, Type messageType, Type interfaceGenericTyp
         static IEnumerable<Type> GetMessageTypesBeingHandledBy(Type type)
         {
             return (from t in type.GetInterfaces()
-                where t.IsGenericType
-                let potentialMessageType = t.GetGenericArguments()[0]
-                where
-                    typeof(IHandleMessages<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t) ||
-                    typeof(IHandleTimeouts<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t)
-                select potentialMessageType)
+                    where t.IsGenericType
+                    let potentialMessageType = t.GetGenericArguments()[0]
+                    where
+                        typeof(IHandleMessages<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t) ||
+                        typeof(IHandleTimeouts<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t)
+                    select potentialMessageType)
                 .Distinct()
                 .ToList();
         }
+        void ValidateHandlerType(Type handlerType)
+        {
+            var propertyTypes = handlerType.GetProperties().Select(p => p.PropertyType).ToList();
+            var ctorArguments = handlerType.GetConstructors()
```

Comment:
But I think we should flatten the hierarchy for properties

https://msdn.microsoft.com/en-us/library/kyaxdd3x(v=vs.110).aspx